modules/substrate/src/verifier.rs

```diff
@@ -258,19 +279,25 @@ where
 		// new authority set change. When we finalize the header we need to update the current
 		// authority set.
 		if header.requires_justification {
+			let signal_hash_proof = "When we import a header we only mark it as
```

Comment:
Can't add comment to unchanged line, so leaving here :)

Could probably be changed as a part of #367 . Now you can change this line above:
```rust
let requires_justification = ancestors.iter().skip(1).find(|h| h.requires_justification);
```

to something that will look read scheduled change from `header.signal_hash` and will check if this change should be enacted by the `header`. So you will only need ancestry to update `is_finalized` flag. Later, when we will add skip-list (or whatever else), we may remove ancestry reads at all => the function would be `O(1)` instead of `O(N)` now.